rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -268,6 +271,219 @@ impl CanisterHistory {
+    /// 3. If deterministic time slicing is disabled, then there are no paused tasks.
+    ///    Aborted tasks may still exist if DTS was disabled in recent checkpoints.
+    pub fn check_dts_invariants(
+        &self,
+        deterministic_time_slicing: FlagStatus,
+        current_round_type: ExecutionRoundType,
+        id: &CanisterId,
+    ) {
+        // There should be at most one paused or aborted task left in the task queue.
+        assert!(
+            self.queue.len() <= 1,
+            "Unexpected tasks left in the task queue of canister {} after a round in canister {:?}",
+            id,
+            self.queue
+        );
+
+        for task in self.queue.iter() {
+            match task {
+                ExecutionTask::AbortedExecution { .. }
+                | ExecutionTask::AbortedInstallCode { .. } => {}
+                ExecutionTask::Heartbeat => {
+                    panic!(
+                        "Unexpected heartbeat task after a round in canister {:?}",
+                        id
+                    );
+                }
+                ExecutionTask::GlobalTimer => {
+                    panic!(
+                        "Unexpected global timer task after a round in canister {:?}",
+                        id
+                    );
+                }
+                ExecutionTask::OnLowWasmMemory(..) => {
+                    panic!(
+                        "Unexpected on low wasm memory task in the queue part of struct TaskQueue, after a round in canister {:?}",
+                        id
+                    );
+                }
+                ExecutionTask::PausedExecution { .. } | ExecutionTask::PausedInstallCode(_) => {
+                    assert_eq!(
+                        current_round_type,
+                        ExecutionRoundType::OrdinaryRound,
+                        "Unexpected paused execution {:?} after a checkpoint round in canister {:?}",
+                        task,
+                        id
+                    );
+
+                    assert_eq!(
+                        deterministic_time_slicing,
+                        FlagStatus::Enabled,
+                        "Unexpected paused execution {:?} with disabled DTS in canister: {:?}",
+                        task,
+                        id
+                    );
+                }
+            }
+        }
+    }
+
+    /// Removes aborted install code task.
+    pub fn remove_aborted_install_code_task(&mut self) {
+        self.queue.retain(|task| match task {
+            ExecutionTask::AbortedInstallCode { .. } => false,
+            ExecutionTask::Heartbeat
+            | ExecutionTask::GlobalTimer
+            | ExecutionTask::OnLowWasmMemory(..)
+            | ExecutionTask::PausedExecution { .. }
+            | ExecutionTask::PausedInstallCode(_)
+            | ExecutionTask::AbortedExecution { .. } => true,
+        });
+    }
+
+    /// Removes aborted `Heartbeat` and `GlobalTimer` tasks.
```

Comment:
```suggestion
    /// Removes `Heartbeat` and `GlobalTimer` tasks.
```

heartbeats and global timers are not "aborted", right?